pkg/disruptors/pod.go

```diff
@@ -104,13 +111,79 @@ func (d *podDisruptor) Targets() ([]string, error) {
 func (d *podDisruptor) InjectHTTPFaults(fault HTTPFault, duration uint, options HTTPDisruptionOptions) error {
 	cmd := buildHTTPFaultCmd(fault, duration, options)
 
-	err := d.controller.ExecCommand(cmd)
+	err := d.portValidator.Validate(fault.Port)
+	if err != nil {
+		return err
+	}
+
+	err = d.controller.ExecCommand(cmd)
 	return err
 }
 
 // InjectGrpcFaults injects faults in the grpc requests sent to the disruptor's targets
 func (d *podDisruptor) InjectGrpcFaults(fault GrpcFault, duration uint, options GrpcDisruptionOptions) error {
 	cmd := buildGrpcFaultCmd(fault, duration, options)
-	err := d.controller.ExecCommand(cmd)
+
+	err := d.portValidator.Validate(fault.Port)
+	if err != nil {
+		return err
+	}
+
+	err = d.controller.ExecCommand(cmd)
 	return err
 }
+
+// PortValidator is responsible for checking whether port is used by every target pod
+type PortValidator struct {
+	ctx       context.Context
+	k8s       kubernetes.Kubernetes
+	namespace string
+	targets   []string
+}
+
+// Validate checks if every target pod listens to target port
+func (pv *PortValidator) Validate(targetPort uint) error {
+	if targetPort == 0 {
+		targetPort = DefaultTargetPort
+	}
+
+	for _, target := range pv.targets {
```

Comment:
Hi @pablochacin ! I think it's almost ready. The last thing that concerns me is that there is a potential performance issue: currently pods are retrieved sequentially in PortValidator. If disruptor will be launched on, let's say, 50 pods, there may be a significant delay. I can parallelise these sequential check with goroutines but the logic probably will be too complex for such trivial task. We can merge it and I'll create a follow-up issue for this or I can try to figure out other options how to overcome this.

Reply:
Yes, I have also noticed this performance issue. I think we better consider options before merging this, because as you said it can have a significant performance penalty.

Reply:
Hi @nickandreev I've been thinking about alternatives to the performance issue. I think the problem comes from making the validation at the wrong place, where the information about the pods is no longer available (only the names).  

The `PodSelector`  has the [logic for retrieving the list of pods in one call](https://github.com/grafana/xk6-disruptor/blob/main/pkg/disruptors/selector.go#L47-L74), but it is not clear to me is how to reuse this logic without breaking encapsulation or creating leaking abstractions. 

Talking about leaking abstractions, this is an issue I have identified in another place: currently the `agentContoller` is [checking if the agent is injected](https://github.com/grafana/xk6-disruptor/blob/main/pkg/disruptors/controller.go#L73-L85) before injecting it. This requires retrieving the Pod using the K8S client and calling the `PodHelper` to inject the agent.  I don't like this code because it breaks the encapsulation of the Kubernetes-related code into the `kubernetes` package and creates some issues for testing the disruptor package.

My intention is to address these issues by [refactoring the Kubernetes-related logic into the `Kubernetes` package](https://github.com/grafana/xk6-disruptor/issues/139).

Maybe after this work is done (hopefully this same week) we can revisit the implementation of the validation.